### DISCUSSION TITLE
How can we return error value from ReflectionFunction?

### DISCUSSION THREAD

**Discussion by exceedsystem** [microsoft/Power-Fx#925] [Q&A] [answered]
Is there a way to return an error from a custom function that returns such as a StringValue?

Example:

```csharp
using Microsoft.PowerFx;
using Microsoft.PowerFx.Types;

var config = new PowerFxConfig();

config.AddFunction(new MyFuncFunction());

var engine = new RecalcEngine(config);

var ret = engine.Eval("If(IsError(MyFunc(\"\")), \"Error\", \"Success\")");

public class MyFuncFunction : ReflectionFunction
{
  public StringValue Execute(StringValue param)
  {
    if (param.Value == "")
    {
      // Failed
      // How to return error from here?
    }
    else
    {
      // Success
      return FormulaValue.New("");
    }
  }
}
```

Best regards.

**Comment by anderson-joyle** [comment, marked as answer]
Check the [CommomErrors](https://github.com/microsoft/Power-Fx/blob/main/src/libraries/Microsoft.PowerFx.Interpreter/Functions/CommonErrors.cs) class. You can call some of the static methods to get a pre-built error message or you can call `CommomErrors.CustomError` to build your own custom error message.
If none of them work for you, it is still a good source to learn how to build ErrorValue objects.

As a reminder, it is also possible to create a ErrorValue from an expression:
`Error({Kind:ErrorKind.InvalidArgument, Message:"This argument is invalid."})`

**Reply by exceedsystem** [reply]
Thanks for the valuable information, I will check it out.

**Reply by exceedsystem** [reply]
I searched the Power Fx source with what you told me and was able to return the error in the following way.

https://github.com/microsoft/Power-Fx/blob/a48e1a7bc710ed0f24adc25802953a849ef3db66/src/tools/Repl/Program.cs#L493

Example:

```csharp
using System;

using Microsoft.PowerFx;
using Microsoft.PowerFx.Types;

var config = new PowerFxConfig();

config.AddFunction(new MyFunc());

var engine = new RecalcEngine(config);

// Has a error
var ret1 = engine.Eval("If(IsError(MyFunc(\"\")), \"Error\", \"Success\")") as StringValue;
Console.WriteLine($"{ret1.Value}");

// No error
var ret2 = engine.Eval("If(IsError(MyFunc(\"Foo\")), \"Error\", \"Success\")") as StringValue;
Console.WriteLine($"{ret2.Value}");

public class MyFunc : ReflectionFunction
{
    // Declare the return type in the base constructor
    public MyFunc()
        : base("MyFunc", FormulaType.String, FormulaType.String)
    {
    }

    // Make the return value a FormulaValue instead of a StringValue
    public FormulaValue Execute(StringValue param)
    {
        if (param.Value.Length == 0)
        {
            //  In case of an error, return an ErrorValue
            return ErrorValue.NewError(new ExpressionError
            {
                Kind = ErrorKind.InvalidArgument,
                Severity = ErrorSeverity.Critical,
                Message = "Invalid parameter."
            });
        }
        else
        {
            // Return value on success
            return FormulaValue.New("");
        }
    }
}
```

Result:

```
Error
Success
```

Thank you so much.

**Reply by anderson-joyle** [reply]
Awesome!

**Reply by MikeStall** [reply]
One nit, errors can be typed. So you should use the 2nd constructor function and pass the type in:

`return ErrorValue.NewError(new ExpressionEror {... },  FormulaValue.String)`

Also curious - can you tell me more about your scenario?

**Reply by exceedsystem** [reply]
@MikeStall
>One nit, errors can be typed. So you should use the 2nd constructor function and pass the type in:
return ErrorValue.NewError(new ExpressionEror {... },  FormulaValue.String)

Thank you for your advice.

I don't have a specific senario yet, as I am still learning and experimenting how to implement Power Fx features in our products.

**Comment by MikeStall** [comment]
FYI - here's an improvement here:  #1018
See example there- now you can just throw a special exception and it will get converted to properly typed error.
